### PR TITLE
retries/sleep for functionaltests

### PR DESCRIPTION
**Pull request by tejasd1990**

### CODE REVIEW

File: node/express/containerCosmosDBWithTests/Tests/functional_tests/sampleFunctionalTests.js [hunk 1]
```diff
@@ -4,6 +4,10 @@ const {until} = require('selenium-webdriver');
 
 process.env['Path'] = process.env['Path'] + ';' + process.env['ChromeWebDriver'];
 
+async function sleep_for_milliseconds(milliseconds){
+	await sleep(milliseconds)
```
Comment (sachinma) [inline node/express/containerCosmosDBWithTests/Tests/functional_tests/sampleFunctionalTests.js]: is this method available in node/javascript

Comment (sachinma) [reply node/express/containerCosmosDBWithTests/Tests/functional_tests/sampleFunctionalTests.js]: Also why do you need the wrapper

Comment (tejasd1990) [reply node/express/containerCosmosDBWithTests/Tests/functional_tests/sampleFunctionalTests.js]: wrapper not needed, removed. added settimeout

File: dotnet/aspnet/webappWithTests/Application/SampleWebApplication.FunctionalTests/SampleFunctionalTests.cs [hunk 2]
```diff
@@ -18,23 +19,43 @@ public static void Initialize(TestContext testContext)
...
-            }
-            finally
+            var numRetries = 5;
+            for (int i = 0; i < numRetries; i++)
             {
-                driver.Quit();
+                try
+                {
+                    var webAppUrl = testContext.Properties["webAppUrl"].ToString();
+                    driver.Navigate().GoToUrl(webAppUrl);
```
Comment (bishal-pdMSFT) [inline dotnet/aspnet/webappWithTests/Application/SampleWebApplication.FunctionalTests/SampleFunctionalTests.cs]: Don't we just need to retry line no 43 and 44.

File: dotnet/aspnet/webappWithTests/Application/SampleWebApplication.FunctionalTests/SampleFunctionalTests.cs [hunk 3]
```diff
@@ -18,23 +19,43 @@ public static void Initialize(TestContext testContext)
...
-                var webAppUrl = testContext.Properties["webAppUrl"].ToString();
-                driver.Navigate().GoToUrl(webAppUrl);
-                Assert.AreEqual("Home Page - My ASP.NET Application", driver.Title, "Expected title to be 'Home Page - My ASP.NET Application'");
-                var filePath = Path.Combine(Path.GetTempPath(), Guid.NewGuid().ToString()) + ".png";
-                var screenshot = driver.GetScreenshot();
-                screenshot.SaveAsFile(filePath);
-                testContext.AddResultFile(filePath);
-            }
-            finally
+            var numRetries = 5;
```
Comment (bishal-pdMSFT) [inline dotnet/aspnet/webappWithTests/Application/SampleWebApplication.FunctionalTests/SampleFunctionalTests.cs]: Should we retry 12 times so as to wait for 1 min

File: java/jsf/webappWithTests/Application/sampleWebApp.FunctionalTests/src/main/java/functional_tests/SampleFunctionalTest.java [hunk 4]
```diff
@@ -27,13 +27,29 @@ public void startBrowser() {
...
+                if (isAlertPresent()) {
+                    System.out.println(isAlertPresent());
+                    driver.switchTo().alert().accept();
+                }
+                driver.get(System.getProperty("webAppUrl"));
+                assertEquals("Sample JSF Application", driver.getTitle());
+                break;
+            }
+            catch(AssertionError e)
+                {
```
Comment (bishal-pdMSFT) [inline java/jsf/webappWithTests/Application/sampleWebApp.FunctionalTests/src/main/java/functional_tests/SampleFunctionalTest.java]: fix indentation

File: node/express/containerCosmosDBWithTests/Tests/functional_tests/sampleFunctionalTests.js [hunk 5]
```diff
@@ -26,13 +26,25 @@ describe('sampleFunctionalTests', function () {
...
-		})
-		.then(() => done())
-		.catch((err) => {
-			done('Failed with error ' + err);
-		});
-    });
+	it('Assert page title', async() => {
+		var numRetries = 5;
+		for (var i = 0; i < numRetries; i++)
+    	{
```
Comment (bishal-pdMSFT) [inline node/express/containerCosmosDBWithTests/Tests/functional_tests/sampleFunctionalTests.js]: fix indentation

File: node/express/containerCosmosDBWithTests/Tests/functional_tests/sampleFunctionalTests.js [hunk 6]
```diff
@@ -26,13 +26,25 @@ describe('sampleFunctionalTests', function () {
...
+			try
+			{
+				await driver.get(process.env['webAppUrl']);
+				await driver.wait(until.titleIs('Express - Node.js Express Application'), 2000);
```
Comment (bishal-pdMSFT) [inline node/express/containerCosmosDBWithTests/Tests/functional_tests/sampleFunctionalTests.js]: there is already a wait until 2 seconds. Would it work if we increase it to 1 min? How does wait work here?

File: dotnet/aspnet/webappWithTests/Application/SampleWebApplication.FunctionalTests/SampleFunctionalTests.cs (hunk 3 above)
Comment (sachinma) [reply dotnet/aspnet/webappWithTests/Application/SampleWebApplication.FunctionalTests/SampleFunctionalTests.cs]: I think the webpage fetch itself is pretty time consuming. So 5 retries should be fine

File: dotnet/aspnet/mssqldbWithTests/Application/SampleWebApplication.FunctionalTests/SampleFunctionalTests.cs [hunk 7]
```diff
@@ -18,23 +19,43 @@ public static void Initialize(TestContext testContext)
...
-            }
-            finally
+            var numRetries = 5;
+            for (int i = 0; i < numRetries; i++)
             {
-                driver.Quit();
+                try
+                {
+                    var webAppUrl = testContext.Properties["webAppUrl"].ToString();
+                    driver.Navigate().GoToUrl(webAppUrl);
```
Comment (sachinma) [inline dotnet/aspnet/mssqldbWithTests/Application/SampleWebApplication.FunctionalTests/SampleFunctionalTests.cs]: Should there be an exception only in this line

**Comment by bishal-pdMSFT** [comment]
@tejasd1990 I and Sachin discussed this and we felt that we should do a time bound retry. Each driver.get() call itself has a time out of 5 mins and hence 5 retries could be too much time.
We can do a larger number of retries (i.e. 12), but check total time elapsed before retry. If more than 10 mins, then fail

File: dotnet/aspnet/webappWithTests/Application/SampleWebApplication.FunctionalTests/SampleFunctionalTests.cs (hunk 3 above)
Comment (tejasd1990) [reply dotnet/aspnet/webappWithTests/Application/SampleWebApplication.FunctionalTests/SampleFunctionalTests.cs]: Added test timeout of 10 mins, with wait time reduced to 2 minutes

File: node/express/containerCosmosDBWithTests/Tests/functional_tests/sampleFunctionalTests.js (hunk 6 above)
Comment (tejasd1990) [reply node/express/containerCosmosDBWithTests/Tests/functional_tests/sampleFunctionalTests.js]: this is the timeout for the title to get populated once the web page is fetched. it is different than the timeout for fetching the webpage

File: dotnet/aspnet/mssqldbWithTests/Application/SampleWebApplication.FunctionalTests/SampleFunctionalTests.cs (hunk 7 above)
Comment (tejasd1990) [reply dotnet/aspnet/mssqldbWithTests/Application/SampleWebApplication.FunctionalTests/SampleFunctionalTests.cs]: Added retry on this line as well as the title check

**Pull request merged**